---
title: "Differential expression analysis FAS with UV treatment"
output: pdf_document
date: "2024-03-03"
---


```{r setup, include=FALSE}
library(data.table)
library(DESeq2)
library(apeglm)
library(ggplot2)
library(ggrepel)
library(EnhancedVolcano)
library(pheatmap)
library(dplyr)
library(PoiClaClu)
library(RColorBrewer)
library(GO.db)
library(GOstats)
library(org.At.tair.db)

#install if necessary
#BiocManager::install(c("AnnotationDbi", "org.At.tair.db"))
```

## Introduction

There are no replicates in the dataset. The replicates named with suffixes _part1 and _part2 are made artificially by dividing the trimmed reads into two groups randomly. They can be viewed as technical replicates. Note that it is not recommended to run DESeq2 on technical replicates and the results are for the exploratory use only, because they are unreliable and are not in alignment with reproducible research. Due to this fact, we focus on genes with high gene counts and set high thresholds to filter data. 

This analysis is based on these tutorials:

- https://ycl6.gitbook.io/guide-to-rna-seq-analysis/differential-expression-analysis/differential-gene-expression/dge-analysis-with-star-+-rsem-input
- https://www.youtube.com/watch?v=wPzeea1Do18&t=421&ab_channel=mikevandewege
- http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#pre-filtering
- https://hbctraining.github.io/DGE_workshop_salmon_online/lessons/04a_design_formulas.html
- https://github.com/hbctraining/DGE_workshop_salmon_online/tree/master/lessons
- https://bioinformatics-core-shared-training.github.io/Bulk_RNAseq_Course_Nov22/Bulk_RNAseq_Course_Base/additional_scripts_and_materials/Statistical_models_in_R_DESeq2.pdf
- https://www.youtube.com/watch?v=X6p3E-QTcUc&t=934s&ab_channel=DnAlabshortreadsequencingworkshop
- https://github.com/tavareshugo/tutorial_DESeq2_contrasts/blob/main/DESeq2_contrasts.md
- https://www.biostars.org/p/9551104/ (discussion on how to filter the results)
- https://www.youtube.com/watch?v=BZyi8L7vBuc&t=76s
- https://davetang.org/muse/2017/08/08/getting-started-arabidopsis-thaliana-genomics/



## Load Count Data 

Load RSEM result files and keep the 5th column containing the expected_count values. Build a countData data.frame to store counts. 
 

```{r}
setwd("C:/Users/kater/DEG_fas_UV/gene_counts_two_parts/rsem")
files = list.files(".", "*genes.results$", full.names = T)

# Loop and read the 5th column remaining files
countData = data.frame(fread(files[1]))[c(1,5)]
for(i in 2:length(files)) {
  countData = cbind(countData, data.frame(fread(files[i]))[5])
}
columns = gsub("Aligned.toTranscriptome.out.rsem.genes.results","",files)
columns = gsub("./","",columns)
colnames(countData) = c("GeneID",columns)
rownames(countData) = countData$GeneID
countData = countData[,c(2:ncol(countData))]
countData = round(countData)    # round to integer

setwd("C:/Users/kater/DEG_fas_UV/gene_counts_two_parts/count_analysis/")


```




## Model Design 

- We run DESeq2 with two models:
  
  + Single factor and 4 levels model. It allows for an easy comparison between groups using constrast function
  + Two factors with interaction model. We use this model to explore interactions or ‘the difference of differences’ by specifying for it in the design formula. For example, if you wanted to explore the effect of genotype on the UV effect, you could specify for it in the design formula as follows ~genotype+treatment+genotype:treatment.Does a treatment have a different effect when given in a different background?

- The analysis is done with a rather high prefiltering cutoff - keeping only genes which have more than 50 reads in all samples. We set alpha (Alpha is the significance cutoff used for optimizing the independent filtering) to 0.05. We then set the adjusted p-value cutoff (FDR) to 0.05 and log2FoldChange cutoff to 1/-1.

- How to interpret log2FoldChange? Fold-change between the comparison and control groups, reported on a logarithmic scale to base 2. E.g. we are testing the “treatment_UV_vs_K” coefficient, positive log2FC indicates a gene up-regulated in the UV treatment compared to the K treatment, while a negative log2FC indicates a down-regulation.

- In summary treatment_group1_vs_group2 means:

  + positive lfc => group1 is up, group2 is down
  + negative lfc => group1 is down, group2 is up
  
### Example pictures




#### Two factors with interaction

![Source: https://bioinformatics-core-shared-training.github.io/Bulk_RNAseq_Course_Nov22/Bulk_RNAseq_Course_Base/additional_scripts_and_materials/Statistical_models_in_R_DESeq2.pdf](images/TwoFactors.png)

#### Single factor and multiple levels model

![Source: https://bioinformatics-core-shared-training.github.io/Bulk_RNAseq_Course_Nov22/Bulk_RNAseq_Course_Base/additional_scripts_and_materials/Statistical_models_in_R_DESeq2.pdf](images/OneFactorMultipleLevels.png)



## Single factor and 4 levels. Testing differential expression

```{r}
## unload info
info_simple <- read.table("info_simple.txt", header = T, sep = '\t')
info_simple

## create DESeqDataSet object 
dds_simple <- DESeqDataSetFromMatrix(countData,info_simple,~group)

## Prefilter genes which have more than 50 reads in all samples.
dds_simple <- estimateSizeFactors(dds_simple)
keep <- rowSums(counts(dds_simple, normalized=TRUE) >= 50 ) >=8
dds_simple <- dds_simple[keep,]

## set the reference level to Wild-type Control
dds_simple$group <- relevel(dds_simple$group, ref = "W_K")

## Run DE
dds_simple <- DESeq(dds_simple)

## Check coefficients for hypothesis testing. Then  we define Contrasts using DESeq coefficient names
resNames_simple <- resultsNames(dds_simple)
resNames_simple

#results(dds, contrast=c("condition","C","B")) meaning genes with logFC > 0 are overexpressed in C.

## Define a function to save the results
extract_res1 <- function(compare_name, contr) {
  
  res = results(dds_simple, alpha = 0.05,  contrast = contr)  

  #Order the table by p-value and export the tables.
  res <- res[order(res$padj),]
  
  # Filtering the results for padj and lfc cutoff
  upregulated = subset(res, padj < 0.05 & log2FoldChange > 1)
  
  # Filtering the results for padj and lfc cutoff
  downregulated = subset(res, padj < 0.05 & log2FoldChange < -1)
  
  summary(upregulated)
  summary(downregulated)
  
  # Save as csv
  write.csv(res, paste(compare_name,"_model1_DESeq2.csv",sep=""))
  write.csv(upregulated, paste(compare_name,"_model1_upregulated_DESeq2.csv",sep=""))
  write.csv(downregulated, paste(compare_name,"_model1_downregulated_DESeq2.csv",sep=""))
  return(res)
}

## See the results
FUV_vs_FK_m1 = extract_res1("FUV_vs_FK", c("group", "F_UV","F_K"))
WUV_vs_WK_m1 = extract_res1("WUV_vs_WK", c("group", "W_UV","W_K"))
FK_vs_WK_m1 = extract_res1("FK_vs_WK", c("group", "F_K","W_K"))
FUV_vs_WUV_m1 = extract_res1("FUV_vs_WUV", c("group", "F_UV","W_UV"))
FUV_vs_WK_m1 = extract_res1("FUV_vs_WK", c("group", "F_UV","W_K"))
```


## Two factors with interaction model. Testing differential expression

- We run DESeq2 with Two Factors Additive model with interaction. It allows for the following comparisons:

  + Interaction "Difference of differences": (W-K vs W-UV) vs (F-K vs F-UV)

- Other available comparison (They are equivalent to the previous model comparisons):

  + F vs W (K)
  + UV vs K (WT)
  + UV vs K (F)
  + F vs W (UV) 





```{r}
## unload info
info <- read.table("info.txt", header = T, sep = '\t')
info

## Create DESeqDataSet object as Two factors with interaction model
dds <- DESeqDataSetFromMatrix(countData,info,~genotype+treatment+genotype:treatment)
dds

## this makes sure the reference level is set as W or K
dds$genotype <- relevel(dds$genotype, ref = "W")
dds$treatment <- relevel(dds$treatment, ref = "K")


# explore model.matrix
model.matrix(design(dds), colData(dds))
```



```{r}
## Define function to filter and save the results
extract_res <- function(compare_name, contr) {
  
  res = results(ddsDE, alpha = 0.05,  contrast = list(contr))  

  #Order the table by p-value and export the tables.
  res <- res[order(res$padj),]
  
  # Filtering the results for padj and lfc cutoff
  upregulated = subset(res, padj < 0.05 & log2FoldChange > 1)
  
  # Filtering the results for padj and lfc cutoff
  downregulated = subset(res, padj < 0.05 & log2FoldChange < -1)
  
  summary(upregulated)
  summary(downregulated)
  
  # Save as csv
  write.csv(res, paste(compare_name,"_model2_DESeq2.csv",sep=""))
  write.csv(upregulated, paste(compare_name,"_model2_upregulated_DESeq2.csv",sep=""))
  write.csv(downregulated, paste(compare_name,"_model2_downregulated_DESeq2.csv",sep=""))
  return(res)
}
```


```{r}
### Prefilter genes which have more than 50 reads in all samples.
dds <- estimateSizeFactors(dds)
keep <- rowSums(counts(dds, normalized=TRUE) >= 50 ) >=8
dds <- dds[keep,]

## running Differential expression analysis
ddsDE <- DESeq(dds)

## Check coefficients for hypothesis testing. Then  we define Contrasts using DESeq coefficient names
resNames <- resultsNames(ddsDE)
resNames
```


### Result: Interaction (W-K vs W-UV) vs (F-K vs F-UV)

Null Hypothesis: Beta3 = 0

```{r}
WKvsWUV_vs_FKvsFUV = extract_res("WKvsWUV_vs_FKvsFUV",c("genotypeF.treatmentUV"))
```


### Treatment UV vs Control (in Wild-types) 

Null Hypothesis: Beta1 = 0

```{r}
UV_vs_K_in_W = extract_res("UV_vs_K_in_W", c("treatment_UV_vs_K")) 

## Show the percentage of all tested genes that have been removed by independent filtering by deseq2. 
## And show the filtering threshold that has been used to filter low count genes 
metadata(UV_vs_K_in_W)$filterThreshold
```

### Fas vs Wild-type (in Controls)

Null Hypothesis: Beta2 = 0

```{r}
F_vs_W_in_K = extract_res("F_vs_W_in_K",c("genotype_F_vs_W"))
```

### Treatment UV vs Control (Fas)
Null Hypothesis: Beta1 + Beta3 = 0

```{r}
UV_Vs_K_in_F = extract_res("UV_Vs_K_in_F", c("treatment_UV_vs_K", "genotypeF.treatmentUV"))
```

### Fas vs Wild-type (UV treated)
Null Hypothesis: Beta2 + Beta3 = 0

```{r}
F_vs_W_in_UV = extract_res("F_vs_W_in_UV",c("genotype_F_vs_W","genotypeF.treatmentUV"))
```

## Export normalized read counts

```{r}
normCounts <- counts(ddsDE,  normalized = T)
write.csv(normCounts, "normalizedReadCounts.csv")

```

## Exploratory plot: heatmap of Poisson distances between samples

Use Poisson distance for raw (non-normalized) count data. Raw reads are better modeled by Poisson distribution than normal distribution.

```{r}
poisd <- PoissonDistance(t(counts(dds)))
samplePoisDistMatrix <- as.matrix(poisd$dd)
rownames(samplePoisDistMatrix) <- paste (dds$sample, sep=" - ")
colnames(samplePoisDistMatrix) <- NULL
colors = colorRampPalette(rev(brewer.pal(9,"Blues")))(255)
pheatmap(samplePoisDistMatrix,clustering_distance_rows = poisd$dd, clustering_distance_cols = poisd$dd,col =colors)
```


## Exploratory plot: PCA plot

The x-axis is the direction that separates the data points the most. The values of the samples in this direction are written PC1. The y-axis is a direction that separates the data the second most. The values of the samples in this direction are written PC2. The percent of the total variance that is contained in the direction is printed in the axis label. Note that these percentages do not add to 100%, because there are more dimensions that contain the remaining variance (although each of these remaining dimensions will explain less than the two that we see).
DESeq2 offers two transformations for count data that stabilize the variance across the mean: the variance stabilizing transformation (VST) for negative binomial data with a dispersion-mean trend (Anders and Huber 2010), implemented in the vst function, and the regularized-logarithm transformation or rlog (Love, Huber, and Anders 2014).

```{r}
vsd <- vst(dds)
plotPCA(vsd, intgroup = "sample")
```
```{r}
rld <- rlog(dds)
plotPCA(rld, intgroup = "sample")
```



## Datasets

Load the upregulated and downregulated genes. We ran DESeq with two models. Model1 gave us these five comparisons:

- FUV_vs_FK
- WUV_vs_WK
- FK_vs_WK
- FUV_vs_WUV
- FUV_vs_WK

Model2 gave us one additional comparison of the difference:

- WKvsWUV_vs_FKvsFUV

```{r}
#Load DESeq results
#1 FUV_vs_FK
setwd("C:/Users/kater/DEG_fas_UV/gene_counts_two_parts/count_analysis/")
FUV_vs_FK_down <- read.csv("FUV_vs_FK_model1_downregulated_DESeq2.csv")
FUV_vs_FK_up<- read.csv("FUV_vs_FK_model1_upregulated_DESeq2.csv")

summary(FUV_vs_FK_up)
summary(FUV_vs_FK_down)

#2 WUV_vs_WK
WUV_vs_WK_down <- read.csv("WUV_vs_WK_model1_downregulated_DESeq2.csv")
WUV_vs_WK_up<- read.csv("WUV_vs_WK_model1_upregulated_DESeq2.csv")

summary(WUV_vs_WK_up)
summary(WUV_vs_WK_down)

#3 FK_vs_WK
FK_vs_WK_down <- read.csv("FK_vs_WK_model1_downregulated_DESeq2.csv")
FK_vs_WK_up<- read.csv("FK_vs_WK_model1_upregulated_DESeq2.csv")

summary(FK_vs_WK_up)
summary(FK_vs_WK_down)

#4 FUV_vs_WUV
FUV_vs_WUV_down <- read.csv("FUV_vs_WUV_model1_downregulated_DESeq2.csv")
FUV_vs_WUV_up<- read.csv("FUV_vs_WUV_model1_upregulated_DESeq2.csv")

summary(FUV_vs_WUV_up)
summary(FUV_vs_WUV_down)

#5 FUV_vs_WK
FUV_vs_WK_down <- read.csv("FUV_vs_WK_model1_downregulated_DESeq2.csv")
FUV_vs_WK_up<- read.csv("FUV_vs_WK_model1_upregulated_DESeq2.csv")

summary(FUV_vs_WK_up)
summary(FUV_vs_WK_down)

#6 WKvsWUV_vs_FKvsFUV
WKvsWUV_vs_FKvsFUV_down <- read.csv("WKvsWUV_vs_FKvsFUV_model2_downregulated_DESeq2.csv")
WKvsWUV_vs_FKvsFUV_up<- read.csv("WKvsWUV_vs_FKvsFUV_model2_upregulated_DESeq2.csv")

summary(WKvsWUV_vs_FKvsFUV_up)
summary(WKvsWUV_vs_FKvsFUV_down)

```

```{r}
# Combine "up" and "down" dataframes into one list
up_down_list <- list(
  FUV_vs_FK_up, WUV_vs_WK_up, FK_vs_WK_up, 
  FUV_vs_WUV_up, FUV_vs_WK_up, WKvsWUV_vs_FKvsFUV_up,
  FUV_vs_FK_down, WUV_vs_WK_down, FK_vs_WK_down, 
  FUV_vs_WUV_down, FUV_vs_WK_down, WKvsWUV_vs_FKvsFUV_down
)

up_down_list_names <- list(
  "FUV_vs_FK_up", "WUV_vs_WK_up", "FK_vs_WK_up", 
  "FUV_vs_WUV_up", "FUV_vs_WK_up", "WKvsWUV_vs_FKvsFUV_up",
  "FUV_vs_FK_down", "WUV_vs_WK_down", "FK_vs_WK_down", 
  "FUV_vs_WUV_down", "FUV_vs_WK_down", "WKvsWUV_vs_FKvsFUV_down"
)

# set names for the list using setNames()
up_down_list <- setNames(up_down_list, up_down_list_names)

# Extract number of observations in each dataframe
num_observations <- sapply(up_down_list, function(df) nrow(df))

# Create a dataframe with the information
data <- data.frame(
  Comparison = rep(c("FUV_vs_FK", "WUV_vs_WK", "FK_vs_WK", 
                     "FUV_vs_WUV", "FUV_vs_WK", "WKvsWUV_vs_FKvsFUV"), 2),
  Direction = rep(c("up", "down"), each = 6),
  Num_Observations = num_observations
)


#Plot using ggplot2 with labels rotated
ggplot(data, aes(x = Comparison, y = Num_Observations, fill = Direction)) +
  geom_bar(stat = "identity", position = "dodge") +
  geom_text(aes(label = Num_Observations), position = position_dodge(width = 0.9), vjust = -0.4) +
  theme_minimal() +
  labs(title = "Number of Upregulated and Downregulated Genes",
       x = "Comparison", y = "Number of Observations") +
  scale_fill_manual(values = c("up" = "blue", "down" = "red")) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

